### PR TITLE
Feat: Add Sentry to batch submitter

### PR DESCRIPTION
**Pull request by annieke**
**Description**
First step in adding alerting to batch submitter and other typescript services. This PR adds Sentry to Batch Submitter, which logs our errors and aggregates them. It is linked to pino-logger which should deliver our error logs to Sentry as well.

**Additional context**
(May also add Sentry to DTL in the same PR)

**Metadata**
https://github.com/ethereum-optimism/roadmap/issues/843

### CODE REVIEW

File: packages/batch-submitter/src/exec/run-batch-submitter.ts [hunk 1]
```diff
@@ -16,7 +17,8 @@ import {
 } from '..'
 
 /* Logger */
-const log = new Logger({ name: 'oe:batch-submitter:init' })
+const destination = createWriteStream({ dsn: process.env.SENTRY_DSN, tracesSampleRate: 1.0 })
```
Comment (snario) [inline packages/batch-submitter/src/exec/run-batch-submitter.ts:20]: We should minimize usage of `process.env` to a single file (i.e., something like `config.ts`)

File: packages/batch-submitter/src/exec/run-batch-submitter.ts [hunk 2]
```diff
@@ -1,5 +1,6 @@
 /* External Imports */
 import { Logger, injectL2Context } from '@eth-optimism/core-utils'
+import { createWriteStream } from 'pino-sentry'
```
Comment (snario) [inline packages/batch-submitter/src/exec/run-batch-submitter.ts:3]: We should maybe make this used in `core-utils`'s `logger.ts`?

File: packages/data-transport-layer/src/services/server/service.ts [hunk 3]
```diff
@@ -101,8 +102,14 @@ export class L1TransportServer extends BaseService<L1TransportServerOptions> {
   private _initializeApp() {
     // TODO: Maybe pass this in as a parameter instead of creating it here?
     this.state.app = express()
+    Sentry.init({
+      dsn: this.options.sentryDsn,
+      tracesSampleRate: 1.0,
+    })
```
Comment (snario) [inline packages/data-transport-layer/src/services/server/service.ts:108]: Does this error if the `dsn` is `undefined`? Do we want tracing?

File: packages/batch-submitter/package.json [hunk 4]
```diff
@@ -31,15 +31,16 @@
...
+    "@eth-optimism/contracts": "^0.2.2",
     "@eth-optimism/core-utils": "^0.2.0",
     "@eth-optimism/ynatm": "^0.2.2",
     "@ethersproject/abstract-provider": "^5.0.5",
     "@ethersproject/providers": "^5.0.14",
     "bluebird": "^3.7.2",
     "dotenv": "^8.2.0",
     "ethers": "5.0.0",
-    "@eth-optimism/contracts": "^0.2.2"
+    "old-contracts": "npm:@eth-optimism/contracts@^0.0.2-alpha.7",
```
Comment (snario) [inline packages/batch-submitter/package.json:42]: Why is this in this PR?

Comment (snario) [reply packages/batch-submitter/package.json:42]: Oh I see, it was moved.

**Review by snario — CHANGES_REQUESTED** [review]
Should we consider making the `destination` a more general purpose thing that `logger.ts` exposes?

File: packages/batch-submitter/src/exec/run-batch-submitter.ts (hunk 1 above)
Comment (annieke) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:20]: `run-batch-submitter` is that file for now; the whole package needs some cleanup for i think for now this is the right place

File: packages/batch-submitter/src/exec/run-batch-submitter.ts (hunk 2 above)
Comment (annieke) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:3]: ^yeah this might be a good idea! pino-sentry uses an older version of pino that i can't consistently overwrite, so i wanted to test out the differences between bs (uses this older version with logging) and dtl (uses the express plugins without logging) during this initial phase, while figuring out the versioning headache

File: packages/data-transport-layer/src/services/server/service.ts (hunk 3 above)
Comment (annieke) [reply packages/data-transport-layer/src/services/server/service.ts:108]: ^i'll test out behavior

tracing is easy to add so can do that! what are your thoughts on tracing one or both these services?

**Comment by tynes** [comment]
We need to ensure that using sentry is optional because we don't want it running with local development

File: packages/data-transport-layer/src/services/server/service.ts (hunk 3 above)
Comment (snario) [reply packages/data-transport-layer/src/services/server/service.ts:108]: I'm not against tracing, it just adds scope. If you think there is a low overhead to adding it then go for it! I think as-is it is sending `1.0` = 100% of requests through the tracing module? Surely that adds overhead? Maybe it should be something like `0.05` = 5%?

Comment (gakonst) [reply packages/data-transport-layer/src/services/server/service.ts:108]: Should probably only initialize if the env bar is set?

File: packages/batch-submitter/src/exec/run-batch-submitter.ts (hunk 2 above)
Comment (annieke) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:3]: seems like this still isn't resolved: https://github.com/yarnpkg/yarn/issues/4874

File: packages/data-transport-layer/src/services/server/service.ts (hunk 3 above)
Comment (annieke) [reply packages/data-transport-layer/src/services/server/service.ts:108]: undefined dsn just issues a warning, so i think we're good here

File: packages/batch-submitter/src/exec/run-batch-submitter.ts (hunk 2 above)
Comment (annieke) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:3]: since i can't update to latest sentry, thoughts on keeping dtl on regular sentry for now and testing out `pino-sentry` through batch submitter? i'm down to tweak this as we go, just don't know what features we might be missing out on yet. i can also make a PR or fork `pino-sentry` to update

Comment (gakonst) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:3]: I think the right approach is to make a PR upstream and update their dep, if not a lot of work?

Comment (annieke) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:3]: sounds good will do!

File: packages/data-transport-layer/src/services/server/service.ts (hunk 3 above)
Comment (annieke) [reply packages/data-transport-layer/src/services/server/service.ts]: @snario added tracing here since it's simple, will monitor and iterate on this

File: packages/batch-submitter/package.json [hunk 5]
```diff
@@ -64,7 +65,8 @@
     "typescript": "^4.2.3"
   },
   "resolutions": {
-    "ganache-core": "^2.13.2"
+    "ganache-core": "^2.13.2",
+    "**/@sentry/node": "^6.2.5"
```
Comment (snario) [inline packages/batch-submitter/package.json:69]: Why do we do this?

File: packages/batch-submitter/src/exec/run-batch-submitter.ts (hunk 1 above)
Comment (snario) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:20]: `tracesSampleRate` at 1.0 is still quite high

File: packages/data-transport-layer/.env.example [hunk 6]
```diff
@@ -22,3 +22,7 @@ DATA_TRANSPORT_LAYER__L2_RPC_ENDPOINT=
 DATA_TRANSPORT_LAYER__TRANSACTIONS_PER_POLLING_INTERVAL=1000
 DATA_TRANSPORT_LAYER__L2_CHAIN_ID=69
 DATA_TRANSPORT_LAYER__LEGACY_SEQUENCER_COMPATIBILITY=false
+
+# Monitoring
+# Leave the SENTRY_DSN variable commented during local development
+# DATA_TRANSPORT_LAYER__SENTRY_DSN=
```
Comment (snario) [inline packages/data-transport-layer/.env.example:28]: Why commented instead of just unset?

File: packages/data-transport-layer/.gitignore [hunk 7]
```diff
@@ -0,0 +1 @@
+db
```
Comment (snario) [inline packages/data-transport-layer/.gitignore:1]: Why is this here? It's not obvious from the commit message

Comment (gakonst) [reply packages/data-transport-layer/.gitignore:1]: Can pull to the top level gitignore

Comment (snario) [reply packages/data-transport-layer/.gitignore:1]: Oh I see, it is due to the DTL creating this folder. I'll add a PR for that.

Comment (snario) [reply packages/data-transport-layer/.gitignore:1]: https://github.com/ethereum-optimism/optimism/pull/487

Comment (annieke) [reply packages/data-transport-layer/.gitignore]: ^updated this in the PR since there was an extra `contracts` in the path

File: packages/batch-submitter/src/exec/run-batch-submitter.ts (hunk 1 above)
Comment (annieke) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts]: ^updated to also be 0.05

**Review by snario — APPROVED** [review]

File: packages/batch-submitter/src/exec/run-batch-submitter.ts [hunk 8]
```diff
@@ -16,7 +17,11 @@ import {
 } from '..'
 
 /* Logger */
-const log = new Logger({ name: 'oe:batch-submitter:init' })
+const destination = createWriteStream({
+  dsn: process.env.SENTRY_DSN,
```
Comment (tynes) [inline packages/batch-submitter/src/exec/run-batch-submitter.ts:21]: What happens if this is not set? We need to make sure this works for local development and CI without sending messages to sentry

Comment (tynes) [reply packages/batch-submitter/src/exec/run-batch-submitter.ts:21]: > So far I've tested that Sentry doesn't report errors without a dsn. We won't have pesky alerts, but I've created an issue here to add environment management so we can do away with running these monitoring services at all during localdev: #491

Sorry I missed this comment, if its not set and its a noop then its all good
